cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -791,27 +905,28 @@ void build_hierarchical(const handle_t& handle,
     RAFT_LOG_TRACE_VEC(fine_clusters_nums.data(), n_mesoclusters);
   }
 
-  auto n_clusters_done = build_fine_clusters(handle,
-                                             n_iters,
-                                             dim,
-                                             dataset,
-                                             mesocluster_labels,
-                                             n_rows,
-                                             fine_clusters_nums.data(),
-                                             fine_clusters_csum.data(),
-                                             mesocluster_sizes,
-                                             n_mesoclusters,
-                                             mesocluster_size_max,
-                                             fine_clusters_nums_max,
-                                             cluster_centers,
-                                             metric,
-                                             &managed_memory,
-                                             device_memory,
-                                             stream);
+  auto n_clusters_done = build_fine_clusters<T, IdxT, LabelT>(handle,
+                                                              n_iters,
+                                                              dim,
+                                                              dataset,
+                                                              dataset_norm,
+                                                              mesocluster_labels,
+                                                              n_rows,
+                                                              fine_clusters_nums.data(),
+                                                              fine_clusters_csum.data(),
+                                                              mesocluster_sizes,
+                                                              n_mesoclusters,
+                                                              mesocluster_size_max,
+                                                              fine_clusters_nums_max,
+                                                              cluster_centers,
+                                                              metric,
+                                                              &managed_memory,
+                                                              device_memory,
+                                                              stream);
   RAFT_EXPECTS(n_clusters_done == n_clusters, "Didn't process all clusters.");
 
   rmm::device_uvector<uint32_t> cluster_sizes(n_clusters, stream, device_memory);
```

Comment:
`cluster_sizes` and, in general, all arrays containing sizes/offsets into the clusters should use the same type as labels, i.e. `LabelT`. Originally, the premise for this module was that the maximum cluster size fits in `uint32_t`; now we assume it can be anything that `LabelT` can represent, right?

Reply:
I don't see why it should be `LabelT`. It would make more sense to use `IdxT` since the cluster size is proportional to the number of samples and inversely proportional to the number of clusters.

But it's also fair to assume that the cluster size will not exceed the range of `uint32_t` and that using a larger type would be inefficient. At the moment, cluster sizes are order of magnitudes below that limit in the worst case.

Reply:
It would almost make sense to have another index type for this, but it might get confusing if we have tons of different index types.

Reply:
Ah, you're right, I mixed these up again. This is the number of clusters what should have the `LabelT` type, not the cluster sizes.